### PR TITLE
Refactor ID handling

### PR DESCRIPTION
**Pull request by jkusa**
- Persists provided ID value when ID field is not `@GeneratedValue`
 - Allows not providing ID value when ID field is `@GeneratedValue`

### CODE REVIEW

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java [hunk 1]
```diff
@@ -544,6 +545,35 @@ public String getId() {
...
+    public void setId(String id) {
+        this.setValue("id", id);
+    }
+
+    /**
+     * Indicates if the ID is generated or not
+     * @return Boolean
+     */
+    public Boolean isIdGenerated() {
+        return getIdAnnotataions().stream().anyMatch(a ->
```
Comment (shadanan) [inline elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: Misspelling.

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java [hunk 2]
```diff
@@ -544,6 +545,35 @@ public String getId() {
...
+                        a.annotationType().equals(GeneratedValue.class)
+        );
+    }
+
+    /**
+     * Returns annotations applied to the ID field
+     * @param value the value
+     * @return Collection of Annotations
+     */
+    private Collection<Annotation> getIdAnnotataions() {
```
Comment (shadanan) [inline elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: Misspelling.

File: elide-core/src/test/java/com/yahoo/elide/core/EntityDictionaryTest.java [hunk 3]
```diff
@@ -58,4 +64,40 @@ public void testGetInverseRelationshipOwnedSide()  {
...
                 "The inverse relationship of children should be parents");
     }
+
+    @Test
+    public void testGetIdAnnotations() throws Exception {
+
+        Collection<Class> expectedAnnotationClasses = Arrays.asList(new Class[]{Id.class, GeneratedValue.class});
+        Collection<Class> actualAnnotationsClasses  = getIdAnnotations(new Parent()).stream()
+            .map(Annotation::annotationType)
+        .collect(Collectors.toList());
```
Comment (shadanan) [inline elide-core/src/test/java/com/yahoo/elide/core/EntityDictionaryTest.java]: indentation

File: elide-core/src/test/java/com/yahoo/elide/core/EntityDictionaryTest.java [hunk 4]
```diff
@@ -58,4 +64,40 @@ public void testGetInverseRelationshipOwnedSide()  {
...
+
+    @Test
+    public void testGetIdAnnotationsSubClass() throws Exception {
+
+        class Friend extends Child { }
+
+        Collection<Class> expectedAnnotationClasses = Arrays.asList(new Class[]{Id.class, GeneratedValue.class});
+        Collection<Class> actualAnnotationsClasses  = getIdAnnotations(new Friend()).stream()
+                .map(Annotation::annotationType)
+        .collect(Collectors.toList());
```
Comment (shadanan) [inline elide-core/src/test/java/com/yahoo/elide/core/EntityDictionaryTest.java]: indentation

File: elide-integration-tests/test/java/com/yahoo/elide/endpoints/ResourceIT.java [hunk 5]
```diff
@@ -1130,6 +1117,53 @@ public void patchExtBadDelete() {
...
+        given()
+            .contentType("application/vnd.api+json")
+            .accept("application/vnd.api+json")
+            .body(req)
+            .post("/assignedId")
+            .then()
+            .statusCode(HttpStatus.SC_FORBIDDEN);
+    }
+
     @Test
```
Comment (shadanan) [inline elide-integration-tests/test/java/com/yahoo/elide/endpoints/ResourceIT.java]: maybe have a test that tries to assign an id to something with the generated annotation?

File: elide-core/src/test/java/com/yahoo/elide/core/EntityDictionaryTest.java (hunk 3 above)
Comment (jkusa) [inline elide-core/src/test/java/com/yahoo/elide/core/EntityDictionaryTest.java]: This style was intentional. Slightly modeled after patterns seen in JS. `stream` and `collect` are the book ends and the `map` is the meat. Are you saying `map` should not have an indent?

Comment (shadanan) [inline elide-core/src/test/java/com/yahoo/elide/core/EntityDictionaryTest.java]: It looked funny to me, but your explanation is reasonable.

File: elide-integration-tests/test/java/com/yahoo/elide/endpoints/ResourceIT.java (hunk 5 above)
Comment (jkusa) [inline elide-integration-tests/test/java/com/yahoo/elide/endpoints/ResourceIT.java]: Most of existing tests do this. Do you think we need another one?

Comment (shadanan) [inline elide-integration-tests/test/java/com/yahoo/elide/endpoints/ResourceIT.java]: Good point.

**Comment by shadanan** [comment]
:+1:

**Comment by yahoocla** [comment]
CLA is valid!

**Pull request merged**